torchvision/transforms/functional_tensor.py

```diff
@@ -651,7 +667,12 @@ def affine(
 
     theta = torch.tensor(matrix, dtype=torch.float).reshape(1, 2, 3)
     shape = img.shape
-    grid = affine_grid(theta, size=(1, shape[-3], shape[-2], shape[-1]), align_corners=False)
+    if shape[-2] == shape[-1]:
+        # here we need normalized translation part of theta
+        grid = affine_grid(theta, size=(1, shape[-3], shape[-2], shape[-1]), align_corners=False)
+    else:
+        # here we need denormalized translation part of theta
+        grid = _gen_affine_grid(theta[0, :, :], w=shape[-1], h=shape[-2], ow=shape[-1], oh=shape[-2])
```

Comment:
I didn't benchmarked both methods to compare the performances, I assumed that `affine_grid` is better optimized that manual `_gen_affine_grid`. That's why I've chosen to split here. Do you mean to wrap everything with `_gen_affine_grid` and split inside this method ?

About an issue in PyTorch, I think https://github.com/pytorch/pytorch/issues/24870 and https://github.com/pytorch/pytorch/issues/36107 already speak about absolute pixel coords. Probably, they are related to this problem.

There is an issue

Reply:
I mean to just dispatch to `_gen_affine_grid`, and not use `nn.functional.affine_grid`. The difference in speed should be very small I think, I'm not sure we dispatch to a cudnn-optimized `affine_grid` so it would be basically the same operations but called from C++